asapdiscovery-data/asapdiscovery/data/openeye.py

```diff
@@ -522,3 +522,289 @@ def openeye_copy_pdb_data(
+    for key, value in data.items():
+        try:
+            key = str(key)
+            value = str(value)
+        except ValueError as v:
+            raise Exception(
+                f"SD data key {key} or value {value} is not castable  a string"
+            ) from v
+        oechem.OESetSDData(mol, key, value)
+    return mol
+
+
+def get_SD_data(mol: oechem.OEMol, key: str) -> str:
+    """
+    Get the SD data on an OpenEye OEMol
+
+    Parameters
+    ----------
+    mol: oechem.OEMol
+        OpenEye OEMol
+
+    Returns
+    -------
+    str
+        SD data value
+    """
+    return oechem.OEGetSDData(mol, key)
+
+
+def get_SD_data_dict(mol: oechem.OEMol) -> dict[str, str]:
+    """
+    Get all SD data on an OpenEye OEMol
+
+    Parameters
+    ----------
+    mol: oechem.OEMol
+        OpenEye OEMol
+
+    Returns
+    -------
+    Dict[str, str]
+        Dictionary of SD data
+    """
+    sd_data = {}
+    for dp in oechem.OEGetSDDataPairs(mol):
+        sd_data[dp.GetTag()] = dp.GetValue()
+    return sd_data
+
+
+def print_SD_Data(mol: oechem.OEMol) -> None:
+    print("SD data of", mol.GetTitle())
+    # loop over SD data
+    for dp in oechem.OEGetSDDataPairs(mol):
+        print(dp.GetTag(), ":", dp.GetValue())
+
+
+def oemol_to_pdb_string(mol: oechem.OEMol) -> str:
+    """
+    Dumps an OpenEye OEMol to a PDB string
+
+    Parameters
+    ----------
+    mol: oechem.OEMol
+         OpenEye OEMol
+
+    Returns
+    -------
+    str
+        PDB string representation of the input OEMol
+    """
+    oms = oechem.oemolostream()
+    oms.SetFormat(oechem.OEFormat_PDB)
+    oms.SetFlavor(oechem.OEFormat_PDB, oechem.OEOFlavor_PDB_Default)
+    oms.openstring()
+    oechem.OEWriteMolecule(oms, mol)
+    molstring = oms.GetString().decode("UTF-8")
+    return molstring
+
+
+def pdb_string_to_oemol(pdb_str: str) -> oechem.OEGraphMol:
+    """
+    Loads a PDB string into an OpenEye OEGraphMol
+
+    Parameters
+    ----------
+    pdb_str: str
+        The string representation of a PDB file
+
+    Returns
+    -------
+    oechem.OEMol
+        resulting OpenEye OEMol
+    """
+    ims = oechem.oemolistream()
+    ims.SetFormat(oechem.OEFormat_PDB)
+    ims.openstring(pdb_str)
+    mols = []
+    mol = oechem.OEGraphMol()
+    for mol in ims.GetOEMols():
+        mols.append(oechem.OEMol(mol))
+    if len(mols) != 1:
+        oechem.OEThrow.Fatal("More than one molecule in input stream")
+    return mols[0]
+
+
+def oedu_to_pdb_string(du: oechem.OEDesignUnit) -> str:
+    """
+    Dumps an OpenEye OEDesignUnit to a PDB string
+
+    Parameters
+    ----------
+    du: oechem.OEDesignUnit
+        OpenEye OEDesignUnit
+
+    Returns
+    -------
+    str
+        PDB string representation of the input OEDesignUnit
+    """
+    oms = oechem.oemolostream()
+    oms.SetFormat(oechem.OEFormat_PDB)
+    oms.openstring()
+    oechem.OEWriteDesignUnit(oms, du)
+    molstring = oms.GetString().decode("UTF-8")
+    return molstring
```

Comment:
What gets written out here? a string representation of the binary OEDU file?

Or just a PDB representation of the protein-ligand system that is in the OEDU?

Can you reconstruct a DU using this string? I don't see any tests yet

Reply:
We have other methods for getting PDB files from DesignUnits and I thought they required a few more hoops to jump through.